test/RelayClient.test.ts

```diff
@@ -1176,5 +1129,403 @@ describe('RelayClient', function () {
         expect(estimation.toString()).to.equal(EXPECTED_ESTIMATION.toString());
       });
     });
+
+    describe('relayTransaction', function () {
+
+      const relayInfo: RelayInfo = {
+        hubInfo: FAKE_HUB_INFO,
+        managerData: {
+          url: 'fake_url'
+        } as RelayManagerData
+      };
+
+      const envelopingRequest = {
+        ...FAKE_RELAY_REQUEST,
+        relayData: {
+          ...FAKE_ENVELOPING_REQUEST_DATA
+        },
+      }
+
+      const transaction = {} as Transaction;
+      let selectNextRelayStub: SinonStub;
+      let attemptRelayTransactionStub: SinonStub;
+      
+      beforeEach(function () {
+        attemptRelayTransactionStub = sandbox.stub().returns(Promise.resolve(transaction));
+        relayClient._getEnvelopingRequestDetails = sandbox.stub().returns(FAKE_RELAY_REQUEST);
+        relayClient._prepareHttpRequest = sandbox.stub();
+        relayClient._verifyEnvelopingRequest = sandbox.stub().returns(Promise.resolve(true));
+        relayClient._attemptRelayTransaction = attemptRelayTransactionStub;
+        selectNextRelayStub = sandbox.stub(utils, 'selectNextRelay').returns(Promise.resolve(relayInfo));
+      });
+
+      afterEach(function () {
+        sandbox.restore();
+      })
+
+      it('should relay the transaction in the first attempt', async function () {
+        const expectedTransaction = await relayClient.relayTransaction(envelopingRequest, FAKE_REQUEST_CONFIG);
+
+        expect(selectNextRelayStub).to.be.calledOnce;
+        expect(expectedTransaction).to.be.equals(transaction);
+      });
+
+      it('should relay the transaction after the first attempt', async function () {
+        attemptRelayTransactionStub.onFirstCall().returns(Promise.resolve(undefined));
+        attemptRelayTransactionStub.onSecondCall().returns(Promise.resolve(transaction));
+        const expectedTransaction = await relayClient.relayTransaction(envelopingRequest, FAKE_REQUEST_CONFIG);
+
+        expect(selectNextRelayStub).to.be.calledTwice;
+        expect(expectedTransaction).to.be.equals(transaction);
+      });
+
+      
+      it('should fail to relay transaction if details are missing', async function () {
+        relayClient._getEnvelopingRequestDetails = sandbox.stub().throws(new Error('missing details'));
+        const expectedTransaction = relayClient.relayTransaction(envelopingRequest, FAKE_REQUEST_CONFIG);
+
+        await expect(expectedTransaction).to.be.rejectedWith('missing details')
+      });
+
+      it('should fail to relay transaction if http request cannot be prepared', async function () {
+        relayClient._prepareHttpRequest = sandbox.stub().throws(new Error('http cannot be prepared'));
```

Comment:
I'd store the error message in a constant

Reply:
this will be changed since Im taking on consideration all the scenarios now

Reply:
updated